### PR TITLE
Fixed receptor collector waiting when everything timeouts

### PR DESCRIPTION
**Pull request by slemrmartin**
Fix of #116. 
When all responses timeout or return error, it ended in a infinite loop

### CODE REVIEW

File: lib/topological_inventory/ansible_tower/receptor/collector.rb [hunk 1]
```diff
@@ -53,12 +52,18 @@ def async_sweep_inventory(refresh_state_uuid, sweep_scope, total_parts, refresh_
         logger.sweeping(:finish, source, sweep_scope, refresh_state_uuid)
       end
 
+      def response_received!
+        self.last_response_at = Time.now.utc
```
Comment (slemrmartin) [inline lib/topological_inventory/ansible_tower/receptor/collector.rb:56]: Although it's not thread safe assignment, in the worst case the value can be different by miliseconds. So it doesn't matter

File: lib/topological_inventory/ansible_tower/receptor/collector.rb [hunk 2]
```diff
@@ -53,12 +52,18 @@ def async_sweep_inventory(refresh_state_uuid, sweep_scope, total_parts, refresh_
         logger.sweeping(:finish, source, sweep_scope, refresh_state_uuid)
       end
 
+      def response_received!
```
Comment (gmcculloug) [inline lib/topological_inventory/ansible_tower/receptor/collector.rb:55]: What is the "!" signifying in the method name?

I know in many Rails methods this is used to say that the method will raise an error if something fails, but it is not clear to me why it is used here.

Comment (slemrmartin) [reply lib/topological_inventory/ansible_tower/receptor/collector.rb:55]: "!" is also used in ruby methods to pointing out that the method is changing internal state of object

File: lib/topological_inventory/ansible_tower/receptor/collector.rb [hunk 3]
```diff
@@ -69,12 +74,18 @@ def wait_for_collected_data
 
         # Wait until all async responses are sent
         while entity_types_collected_cnt.value < service_catalog_entity_types.size
-          if last_save_at.nil? || (last_save_at < ReceptorController::Client::Configuration.default.response_timeout.minutes.ago && last_save_at >= max_wait_sync_threshold.minutes.ago)
+          if last_response_at.nil? && refresh_started_at >= max_wait_sync_threshold.minutes.ago
+            logger.warn("[ASYNC] Collector for source_uid: #{source}: No response received since #{refresh_started_at}")
+            sleep(60)
+          elsif last_response_at < ReceptorController::Client::Configuration.default.response_timeout.minutes.ago && last_response_at >= max_wait_sync_threshold.minutes.ago
```
Comment (gmcculloug) [inline lib/topological_inventory/ansible_tower/receptor/collector.rb]: Will this condition throw an error if `last_response_at` is `nil`?  (Same for the last `elsif` block)

```ruby
nil < 100
=> NoMethodError (undefined method `<' for nil:NilClass)
```

Should there be a higher level condition here that checks for `if last_response_at` and then do further processing inside of that?

```ruby
while entity_types_collected_cnt.value < service_catalog_entity_types.size
  if last_response_at
    ...
  else
    ...
  end
end
```

Comment (slemrmartin) [reply lib/topological_inventory/ansible_tower/receptor/collector.rb]: ou, you're right

**Review by roliveri — APPROVED** [review]

**Review by syncrou — APPROVED** [review]
LGTM 👍

**Pull request merged**